### PR TITLE
Add fused add_relu op.

### PR DESCRIPTION
**Pull request by kimishpatel**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40252 Add add_relu fusion pass to optimize_for_mobile.
* #39343 JIT pass for add relu fusion.
* **#39342 Add fused add_relu op.**
* #39341 [Vec256][neon] Add neon backend for vec256
* #40059 Add benchmark for add op.

Summary:
Many networks such as resnet have adds followed by relu. This op is the
first step in enabling this fused implementation.
Once we have the fused add_relu op, a JIT pass will be written to
replace add + relu patterns with add_relu.

Test Plan:
python test/test_nn.py TestAddRelu

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21822397](https://our.internmc.facebook.com/intern/diff/D21822397)

### CODE REVIEW

File: aten/src/ATen/native/cpu/BinaryOpsKernel.cpp [hunk 1]
```diff
@@ -35,6 +35,27 @@ void add_kernel(TensorIterator& iter, Scalar alpha_scalar) {
   }
 }
 
+void add_clamp_kernel(TensorIterator& iter, Scalar alpha_scalar, Scalar min_val, Scalar max_val) {
+  AT_DISPATCH_ALL_TYPES(iter.dtype(), "add_relu_cpu", [&]() {
```
Comment (dreiss) [inline aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:39]: "add_clamp_cpu"?

Comment (kimishpatel) [reply aten/src/ATen/native/cpu/BinaryOpsKernel.cpp:39]: Yes. Will fix it.

**Comment by dreiss** [comment]
Do you have any numbers on how much this improves resnet performance compared to unfused (but vec256-ized) `add` then `relu`?

**Comment by dreiss** [comment]
(or any other model like unet)

**Comment by kimishpatel** [comment]
> Do you have any numbers on how much this improves resnet performance compared to unfused (but vec256-ized) `add` then `relu`?

I havent tried on resnet. For some reason on unet it did not make any difference. I did not dig further yet to find out why. I will try it out on resnet.

**Comment by kimishpatel** [comment]
@dreiss, I checked perf on unet model on s9 phone.
Before this stack: 45ms
After this stack and applying add_relu fusion pass (since I am yet to add this to optimizeForMobile, will stack another diff on top for that): 33ms
As a reference just with add_relu fusion and no neon backend we get 37ms.

**Review by dreiss — APPROVED** [review]

**Review by dreiss — APPROVED** [review]

**Pull request closed without merging**